src/daemon/db_util.rs

```diff
@@ -0,0 +1,343 @@
+        ts
+    } else {
+        let car_stream = CarStream::new(tokio::io::BufReader::new(
+            tokio::fs::File::open(&downloaded_car_temp_path).await?,
+        ))
+        .await?;
+        let roots = car_stream.header.roots.clone();
+        // Use another temp file to make sure all final `.forest.car.zst` files are complete and valid.
+        let forest_car_db_temp_path =
+            tempfile::NamedTempFile::new_in(forest_car_db_dir)?.into_temp_path();
+        {
+            let mut writer =
+                tokio::io::BufWriter::new(tokio::fs::File::create(&forest_car_db_temp_path).await?);
+            let frames = crate::db::car::forest::Encoder::compress_stream(
+                8000usize.next_power_of_two(),
+                zstd::DEFAULT_COMPRESSION_LEVEL as _,
+                car_stream.map_err(anyhow::Error::from),
+            );
+            crate::db::car::forest::Encoder::write(&mut writer, roots, frames).await?;
+            writer.shutdown().await?;
+        }
+        let ts =
+            ForestCar::new(RandomAccessFile::open(&forest_car_db_temp_path)?)?.heaviest_tipset()?;
+        forest_car_db_temp_path.persist(&forest_car_db_path)?;
+        ts
+    };
+
+    info!(
+        "Imported snapshot in: {}s, tipset epoch: {}",
+        stopwatch.elapsed().as_secs(),
+        ts.epoch()
+    );
+
+    Ok((forest_car_db_path, ts))
+}
+
+/// If our current chain is below a supported height, we need a snapshot to bring it up
+/// to a supported height. If we've not been given a snapshot by the user, get one.
+///
+/// An [`Err`] should be considered fatal.
+async fn fetch_snapshot_if_required(
+    config: &mut Config,
+    epoch: ChainEpoch,
+    auto_download_snapshot: bool,
+    download_directory: &Path,
+) -> anyhow::Result<()> {
+    if !download_directory.is_dir() {
+        anyhow::bail!(
+            "`download_directory` does not exist: {}",
+            download_directory.display()
+        );
+    }
+
+    let vendor = snapshot::TrustedVendor::default();
+    let chain = &config.chain.network;
+
+    // What height is our chain at right now, and what network version does that correspond to?
+    let network_version = config.chain.network_version(epoch);
+    let network_version_is_small = network_version < NetworkVersion::V16;
+
+    // We don't support small network versions (we can't validate from e.g genesis).
+    // So we need a snapshot (which will be from a recent network version)
+    let require_a_snapshot = network_version_is_small;
+    let have_a_snapshot = config.client.snapshot_path.is_some();
+
+    match (require_a_snapshot, have_a_snapshot, auto_download_snapshot) {
+        (false, _, _) => Ok(()),   // noop - don't need a snapshot
+        (true, true, _) => Ok(()), // noop - we need a snapshot, and we have one
+        (true, false, true) => {
+            // we need a snapshot, don't have one, and have permission to download one, so do that
+            let max_retries = 3;
```

Comment:
We retry three times here but don't do any retries in the other branch? Let's refactor to remove the duplicate code.

Reply:
I can refactor this with a separate PR, as the code is not part of this PR (only moved this function)

Reply:
Okay, let's move it back to the original location, then. If you don't touch the code, you don't have to fix it. If you touch the code, you fix it. :)

Reply:
Moved back, the retry lines are no longer part of diff